perf/impl/go-libp2p/v0.27/main.go

```diff
@@ -0,0 +1,119 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/libp2p/go-libp2p"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/multiformats/go-multiaddr"
+)
+
+func main() {
+	runServer := flag.Bool("run-server", false, "Should run as server")
+	serverAddr := flag.String("server-address", "", "Server address")
+	transport := flag.String("transport", "tcp", "Transport to use")
+	uploadBytes := flag.Uint64("upload-bytes", 0, "Upload bytes")
+	downloadBytes := flag.Uint64("download-bytes", 0, "Download bytes")
+	flag.Parse()
+
+	host, port, err := net.SplitHostPort(*serverAddr)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	tcpMultiAddrStr := fmt.Sprintf("/ip4/%s/tcp/%s", host, port)
+	quicMultiAddrStr := fmt.Sprintf("/ip4/%s/udp/%s/quic-v1", host, port)
+
+	var opts []libp2p.Option
+	if *runServer {
+		opts = append(opts, libp2p.ListenAddrStrings(tcpMultiAddrStr, quicMultiAddrStr))
+
+		// Generate stable fake identity.
+		//
+		// Using a stable identity (i.e. peer ID) allows the client to
+		// connect to the server without a prior exchange of the
+		// server's peer ID.
+		priv, _, err := crypto.GenerateEd25519Key(&simpleReader{seed: 0})
+		if err != nil {
+			log.Fatalf("failed to generate key: %s", err)
+		}
+		opts = append(opts, libp2p.Identity(priv))
+	}
+
+	h, err := libp2p.New(opts...)
+	if err != nil {
+		log.Fatalf("failed to instantiate libp2p: %s", err)
+	}
+
+	perf := NewPerfService(h)
+	if *runServer {
+		for _, a := range h.Addrs() {
+			fmt.Println(a.Encapsulate(multiaddr.StringCast("/p2p/" + h.ID().String())))
+		}
+
+		select {} // run forever, exit on interrupt
+	}
+
+	var multiAddrStr string
+	switch *transport {
+	case "tcp":
+		multiAddrStr = tcpMultiAddrStr
+	case "quic-v1":
+		multiAddrStr = quicMultiAddrStr
+	default:
+		fmt.Println("Invalid transport. Accepted values: 'tcp' or 'quic-v1'")
+		return
+	}
+	// Peer ID corresponds to the above fake identity.
+	multiAddrStr = multiAddrStr + "/p2p/12D3KooWDpJ7As7BWAwRMfu1VU2WCqNjvq387JEYKDBj4kx6nXTN"
```

Comment:
I think this should be passed in as part of the multiaddr for the server, I don't see the benefit of hard coding this identity. We are already printing it anyways. And then you can remove all the code around generating keys from a seed.

Reply:
> We are already printing it anyways.

I am not sure which place you are referring to.

Today there is no way for a perf `impl/` server deployment to pass information (e.g. its peer ID) to its perf `impl/` client deployment counterpart.

We can add a mechanism, e.g. reading the first stdout line of the server or reading the identity from a file.

Given that we are benchmarking performance here and not testing the crypto handshakes, I think this hack is fine, i.e. I don't think it is worth it adding the additional mechanism. In case you think differently @MarcoPolo I suggest doing this as a follow-up.

> I think this should be passed in as part of the multiaddr for the server

Note that the `runner/` is no longer aware of the concept of a multiaddr. It only provides IP and port to the server and client. Otherwise we would need to partse multiaddr in `impl/quic-go` and `impl/https`.